File: ray-tracing-web/src/source/surfaces/Checkerboard.ts
import { Color } from '../Color';
import { Vector } from '../Vector';
import { Surface } from '../Surface';

export class Checkerboard implements Surface {
    public readonly roughness = 150

    public diffuse(pos: Vector): Color {
        if ((Math.floor(pos.z) + Math.floor(pos.x)) % 2 !== 0) {
            return Color.white;
        }

        return Color.black;

    }

    public specular(pos: Vector): Color {
        return Color.white;
    }

    public reflect(pos: Vector): number {
        if ((Math.floor(pos.z) + Math.floor(pos.x)) % 2 !== 0) {
            return 0.1;
        }

        return 0.7;
    }
}

File: ray-tracer/elements/surfaces/Checkerboard.cs
using System;
using ray_tracer.common;
using ray_tracer.scene;


namespace ray_tracer.elements.surfaces
{
    public class Checkerboard : Surface
    {
        public double roughness => 150.0;

        public Color diffuse(Vector pos)
        {
            if (isBlackField(pos))
            {
                return Color.white;
            }

            return Color.black;
        }

        public double reflect(Vector pos)
        {
            if (isBlackField(pos))
            {
                return 0.1;
            }

            return 0.7;
        }

        public Color specular(Vector pos)
        {
            return Color.white;
        }

        private bool isBlackField(Vector pos)
        {
            return (int) (Math.Floor(pos.z) + Math.Floor(pos.x)) % 2 != 0;
        }
    }
}
